transmission/transmission.go

```diff
@@ -84,57 +86,97 @@ func (h *Honeycomb) Start() error {
 	}
 	h.Logger.Printf("default transmission starting")
 	h.responses = make(chan Response, h.PendingWorkCapacity*2)
-	h.muster.MaxBatchSize = h.MaxBatchSize
-	h.muster.BatchTimeout = h.BatchTimeout
-	h.muster.MaxConcurrentBatches = h.MaxConcurrentBatches
-	h.muster.PendingWorkCapacity = h.PendingWorkCapacity
 	if h.Metrics == nil {
 		h.Metrics = &nullMetrics{}
 	}
-	h.muster.BatchMaker = func() muster.Batch {
-		return &batchAgg{
-			userAgentAddition: h.UserAgentAddition,
-			batches:           map[string][]*Event{},
-			httpClient: &http.Client{
-				Transport: h.Transport,
-				Timeout:   60 * time.Second,
-			},
-			blockOnResponse:       h.BlockOnResponse,
-			responses:             h.responses,
-			metrics:               h.Metrics,
-			disableCompression:    h.DisableGzipCompression || h.DisableCompression,
-			enableMsgpackEncoding: h.EnableMsgpackEncoding,
+	if h.batchMaker == nil {
+		h.batchMaker = func() muster.Batch {
+			return &batchAgg{
+				userAgentAddition: h.UserAgentAddition,
+				batches:           map[string][]*Event{},
+				httpClient: &http.Client{
+					Transport: h.Transport,
+					Timeout:   60 * time.Second,
+				},
+				blockOnResponse:       h.BlockOnResponse,
+				responses:             h.responses,
+				metrics:               h.Metrics,
+				disableCompression:    h.DisableGzipCompression || h.DisableCompression,
+				enableMsgpackEncoding: h.EnableMsgpackEncoding,
+			}
 		}
 	}
+
+	m := h.createMuster()
+	h.muster = m
 	return h.muster.Start()
 }
 
+func (h *Honeycomb) createMuster() *muster.Client {
+	m := new(muster.Client)
+	m.MaxBatchSize = h.MaxBatchSize
+	m.BatchTimeout = h.BatchTimeout
+	m.MaxConcurrentBatches = h.MaxConcurrentBatches
+	m.PendingWorkCapacity = h.PendingWorkCapacity
+	m.BatchMaker = h.batchMaker
+	return m
+}
+
 func (h *Honeycomb) Stop() error {
 	h.Logger.Printf("Honeycomb transmission stopping")
 	err := h.muster.Stop()
 	close(h.responses)
 	return err
 }
 
+func (h *Honeycomb) Flush() (err error) {
+	m := h.muster
+	defer func() {
+		e := m.Stop() // This is latent, so we do not want to execute this with the lock.
+		if err == nil {
+			err = e
+			// bubble up the error from Stop if there was no error from Start
+		}
+	}()
+	h.musterLock.Lock()
+	defer h.musterLock.Unlock()
+	newMuster := h.createMuster()
+	h.muster = newMuster
+	return h.muster.Start()
+}
+
 func (h *Honeycomb) Add(ev *Event) {
+	added := h.tryAdd(ev)
+	if added {
+		h.Metrics.Increment("messages_queued")
+		return
+	}
```

Comment:
I do like using named variables to clarify the flow - in this case though the `tryAdd` seems clear enough to skip it.
```suggestion
	if h.tryAdd(ev) {
		h.Metrics.Increment("messages_queued")
		return
	}
```